### PR TITLE
[SC-477] Add tests for Issue/Reissue/Burn via @Callable

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+  "Restrictions in @Callable" - {
+
+    "Issue 10 assets should not produce an error" in {
+      val acc = createDapp(script(simpleNonreissuableAsset))
+      val tx  = invokeScript(acc, "issue10Assets")
+
+      for (nth <- 0 to 9) validateIssuedAssets(acc, tx, simpleNonreissuableAsset, nth, "@Callable")
+    }
+
+    "Issue more then 10 assets should produce an error" in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: than

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+      val txIssue = invokeScript(acc, "issueAsset")
+      val assetId = validateIssuedAssets(acc, txIssue, simpleNonreissuableAsset, method = method)
+      assertBadRequestAndMessage(
+        reissue(acc, method, assetId, 100500, false),
+        "State check failed. Reason: Asset is not reissuable"
+      )
+    }
+  }
+
+  "Restrictions in @Callable" - {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: What if two assets are issued with the same nonce

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+      val txIssue = invokeScript(acc, "issueAsset", fee = fee)
+      val assetId = validateIssuedAssets(acc, txIssue, data, method = method)
+
+      burn(acc, method, assetId, data("quantity").asInstanceOf[Number].intValue)
+
+      sender.assetsDetails(assetId).quantity shouldBe 0
+      sender.assertAssetBalance(acc, assetId, 0)
+    }
+
+    "Reissuable asset could be reissued" in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: what if the resulting quantity is greater than Long.MAX_VALUE?

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+      val initialReissuable = simpleReissuableAsset("isReissuable").asInstanceOf[Boolean].booleanValue
+
+      reissue(acc, method, assetId, addedQuantity, !initialReissuable)
+
+      sender.assetsDetails(assetId).reissuable shouldBe !initialReissuable
+      sender.assetsDetails(assetId).quantity shouldBe initialQuantity + addedQuantity
+      sender.assertAssetBalance(acc, assetId, initialQuantity + addedQuantity)
+    }
+
+    "Non-reissuable asset could not be reissued" in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: What if try to reissue NFT?

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 3 above)
Comment (msmolyakov) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: What if try to reissue/burn a smart asset that restricts such actions?

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  val setScriptPrice      = 0.01.waves
+
+  def script(asset: Map[String, Any], function: String = ""): String = {
+    val s           = if (asset("compiledScript").equals("")) "unit" else asset("compiledScript")
+    val d           = asset("decimals")
+    val ds          = asset("description")
+    val r           = asset("isReissuable")
+    val n           = asset("name")
+    val q           = asset("quantity")
+    val ns          = asset("nounce")
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: ok

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 1 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: ok

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 2 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: ok

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 3 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: what if the resulting quantity is greater than Long.MAX_VALUE? -- error, I've added the test

Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: What if try to reissue/burn a smart asset that restricts such actions? -- it should produce an error

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 4 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: error, I've added the test

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        reissue(acc, method, assetId, 100500, false),
+        "Asset is not reissuable"
+      )
+    }
+  }
+
+  "Restrictions in @Callable" - {
+    val method = "@Callable"
+
+    "Issue two identical with the same nonce should not produce an error " in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: Try to issue both assets in one invocation. It must produce error

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+      val assetId         = validateIssuedAssets(acc, txIssue, simpleReissuableAsset, method = method)
+      val initialQuantity = simpleReissuableAsset.q
+
+      assertBadRequestAndMessage(
+        reissue(acc, method, assetId, Long.MaxValue - initialQuantity + 1, true),
+        "State check failed. Reason: Asset total value overflow"
+      )
+    }
+
+    "Burning more than current asset count should produce an error" in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: Try to transfer part of tokens to another account before burning here.

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    "Burn to 1, set reissuee false, reissue" in {}
+
+    "Issue 10 assets should not produce an error" in {
+      val acc = createDapp(script(simpleNonreissuableAsset))
+      val tx  = invokeScript(acc, "issue10Assets")
+
+      for (nth <- 0 to 9) validateIssuedAssets(acc, tx, simpleNonreissuableAsset, nth, "@Callable")
+    }
+
+    "Issue more than 10 assets should produce an error" in {
```
Comment (msmolyakov) [inline node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: Check the same for Reissue and Burn separately and together

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 7 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: there is a bug here (SC-576)

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 6 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: bug SC-575

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 8 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: ok

File: node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala (hunk 7 above)
Comment (ngadiyak) [reply node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala]: +

**Pull request merged**